packages/protobuf/src/transform/index.ts

```diff
@@ -0,0 +1,933 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import {
+  DiagnosticTarget,
+  Enum,
+  getEffectiveModelType,
+  getTypeName,
+  Interface,
+  isDeclaredInNamespace,
+  Model,
+  ModelProperty,
+  Namespace,
+  Operation,
+  Program,
+  resolvePath,
+  Scalar,
+  StringLiteral,
+  SyntaxKind,
+  Type,
+  Union,
+} from "@typespec/compiler";
+import {
+  map,
+  matchType,
+  ProtoEnumDeclaration,
+  ProtoFieldDeclaration,
+  ProtoFile,
+  ProtoMap,
+  ProtoMessageBodyDeclaration,
+  ProtoMessageDeclaration,
+  ProtoMethodDeclaration,
+  ProtoOneOfDeclaration,
+  ProtoRef,
+  ProtoScalar,
+  ProtoTopLevelDeclaration,
+  ProtoType,
+  ref,
+  scalar,
+  ScalarIntegralName,
+  StreamingMode,
+  unreachable,
+} from "../ast.js";
+import { ProtobufEmitterOptions, reportDiagnostic, state } from "../lib.js";
+import { $field, isMap, Reservation } from "../proto.js";
+import { writeProtoFile } from "../write.js";
+
+// Cache for scalar -> ProtoScalar map
+const _protoScalarsMap = new WeakMap<Program, Map<Type, ProtoScalar>>();
+
+/**
+ * Create a worker function that converts the TypeSpec program to Protobuf and writes it to the file system.
+ */
+export function createProtobufEmitter(
+  program: Program
+): (outDir: string, options: ProtobufEmitterOptions) => Promise<void> {
+  return async function doEmit(outDir, options) {
+    // Convert the program to a set of proto files.
+    const files = tspToProto(program);
+
+    if (!program.compilerOptions.noEmit && !options?.noEmit && !program.hasError()) {
+      for (const file of files) {
+        // If the file has a package, emit it to a path that is shaped like the package name. Otherwise emit to
+        // main.proto
+
+        // Collisions have already been detected.
+
+        const packageSlug = file.package?.split(".") ?? ["main"];
+        const filePath = resolvePath(outDir, ...packageSlug.slice(0, -1));
+
+        await program.host.mkdirp(filePath);
+        await program.host.writeFile(
+          resolvePath(filePath, packageSlug[packageSlug.length - 1] + ".proto"),
+          writeProtoFile(file)
+        );
+      }
+    }
+  };
+}
+
+/**
+ * Create a set of proto files that represent the TypeSpec program.
+ *
+ * This is the meat of the emitter.
+ */
+function tspToProto(program: Program): ProtoFile[] {
+  const packages = new Set<Namespace>(
+    program.stateMap(state.package).keys() as Iterable<Namespace>
```

Comment:
I think having an accessor for each one of those stateMap read would be nice so you don't have to have `as` in the logic and potentially have to keep casting it multiple times.

Reply:
I agree this would be nice. I added a tracking issue for some refactoring work here: https://github.com/microsoft/typespec/issues/1855